File: backend/app/schemas/audit_log.py
import uuid
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, Field


class AuditLogResponse(BaseModel):
    """Pydantic schema representing audit log retrieval query results."""

    id: uuid.UUID
    user_id: Optional[uuid.UUID] = None
    action: str
    entity_type: Optional[str] = None
    entity_id: Optional[str] = None
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None
    timestamp: datetime
    # Map from SQLAlchemy 'meta_data' attribute to 'metadata' JSON key
    metadata: Optional[dict] = Field(None, validation_alias="meta_data")

    model_config = {
        "from_attributes": True,
        "populate_by_name": True,
    }

File: frontend/lib/features/audit_logs/domain/audit_log.dart
class AuditLog {
  final String id;
  final String? userId;
  final String action;
  final String? entityType;
  final String? entityId;
  final String? ipAddress;
  final String? userAgent;
  final DateTime timestamp;
  final Map<String, dynamic>? metadata;

  const AuditLog({
    required this.id,
    this.userId,
    required this.action,
    this.entityType,
    this.entityId,
    this.ipAddress,
    this.userAgent,
    required this.timestamp,
    this.metadata,
  });

  factory AuditLog.fromJson(Map<String, dynamic> json) {
    return AuditLog(
      id: json['id'] as String,
      userId: json['user_id'] as String?,
      action: json['action'] as String? ?? '',
      entityType: json['entity_type'] as String?,
      entityId: json['entity_id'] as String?,
      ipAddress: json['ip_address'] as String?,
      userAgent: json['user_agent'] as String?,
      timestamp: DateTime.parse(json['timestamp'] as String),
      metadata: json['metadata'] as Map<String, dynamic>?,
    );
  }
}

class PaginatedAuditLogs {
  final List<AuditLog> items;
  final int total;
  final int skip;
  final int limit;

  const PaginatedAuditLogs({
    required this.items,
    required this.total,
    required this.skip,
    required this.limit,
  });

  factory PaginatedAuditLogs.fromJson(Map<String, dynamic> json) {
    final items = (json['items'] as List<dynamic>? ?? [])
        .map((e) => AuditLog.fromJson(e as Map<String, dynamic>))
        .toList();
    return PaginatedAuditLogs(
      items: items,
      total: json['total'] as int? ?? items.length,
      skip: json['skip'] as int? ?? 0,
      limit: json['limit'] as int? ?? items.length,
    );
  }
}
